rust/noosphere-storage/src/migration.rs

```diff
@@ -0,0 +1,234 @@
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait ExportStorage
+where
+    Self: Storage,
+    <Self as Storage>::KeyValueStore: IterableStore,
+{
+    async fn get_all_store_names(&self) -> Result<Vec<String>> {
+        let mut names = vec![];
+        names.extend(SPHERE_DB_STORE_NAMES.iter().map(|name| String::from(*name)));
+        Ok(names)
+    }
+}
+
+impl<S> ExportStorage for S
+where
+    S: Storage,
+    S::KeyValueStore: IterableStore,
+{
+}
+
+/// A blanket implementation for [Storage]s to import
+/// an [ExportStorage] storage.
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait ImportStorage<'a, E>
+where
+    Self: Storage,
+    Self::KeyValueStore: Store,
+    E: ExportStorage + ConditionalSync + 'a,
+    <E as Storage>::KeyValueStore: IterableStore,
+{
+    /// Copy all stores' entries from `exportable` into this [Storage].
+    async fn import(&'a mut self, exportable: &E) -> Result<()> {
+        for store_name in exportable.get_all_store_names().await? {
+            let mut store = self.get_key_value_store(&store_name).await?;
+            let export_store = exportable.get_key_value_store(&store_name).await?;
+            let mut stream = export_store.get_all_entries();
+            while let Some((key, value)) = stream.try_next().await? {
+                if let Some(value) = value {
+                    Store::write(&mut store, key.as_ref(), value.as_ref()).await?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+impl<'a, T, E> ImportStorage<'a, E> for T
+where
+    T: Storage,
+    T::KeyValueStore: Store,
+    E: ExportStorage + ConditionalSync + 'a,
+    <E as Storage>::KeyValueStore: IterableStore,
+{
+}
+
+/// A [Storage] that can be used as the source in [migrate_storage].
+pub trait MigratableSource
+where
+    Self: OpenStorage + RemoveStorage,
+    <Self as Storage>::KeyValueStore: IterableStore,
+{
+}
+impl<T> MigratableSource for T
+where
+    Self: OpenStorage + RemoveStorage,
+    <Self as Storage>::KeyValueStore: IterableStore,
+{
+}
+
+/// A [Storage] that can be used as the destination target in [migrate_storage].
+pub trait MigratableTarget<'a, S>
+where
+    S: MigratableSource + ConditionalSync + 'a,
+    <S as Storage>::KeyValueStore: IterableStore,
+    Self: ImportStorage<'a, S> + Storage + OpenStorage + RemoveStorage + RenameStorage,
+    <Self as Storage>::KeyValueStore: Store,
+{
+}
+
+impl<'a, S, T> MigratableTarget<'a, S> for T
+where
+    S: MigratableSource + ConditionalSync + 'a,
+    <S as Storage>::KeyValueStore: IterableStore,
+    Self: ImportStorage<'a, S> + Storage + OpenStorage + RemoveStorage + RenameStorage,
+    <Self as Storage>::KeyValueStore: Store,
```

Comment:
Unless `<Self as Storage>::KeyValueStore: Store` and ` <S as Storage>::KeyValueStore: IterableStore` can be inherited by a supertrait, `MigratableTarget` and `MigratableSource` can probably be folded away